### ISSUE TITLE
How to train custom dataset on BERT-GT model

### ISSUE THREAD

**Issue by ArnavHP18** [ncbi/bert_gt#1]
We have our dataset in which we have annotated the entities, relations, and coreferences. So how do we train a custom dataset such that it can be implemented using the bert-gt model?

**Comment by ptlai** [comment]
Hello ArnavHP18,

Sorry for the late response to you. Thank you for your interest in our work.
To train and test the bert-gt on a custom dataset, there are two steps:

1. To convert your dataset into bert-gt's format files, including train.tsv, dev.tsv, and test.tsv. Our codes (in src/dataset_format_converter/) only provide scripts for converting ChemProt/DDI/CDR/n-ary datasets into bert-gt format.
So if you can convert your dataset into one of their formats, you can use the corresponding script to convert your dataset into the bert-gt format files.
Alternatively, if your dataset cannot convert to one of their formats. You might have to implement your converter to convert your dataset into the bert-gt format file. You should prepare three bert-gt format files (train.tsv, dev.tsv, and test.tsv)
You can find the examples of ChemProt/DDI/CDR/n-ary datasets by using our scripts. In general, each file has at least four columns, including 'index,' 'text,' 'neighbors,' 'label.' If you have any problem with the input files, please let me know.

2. You have to implement your own YourDatasetProcessor in the run_bert_gt.py as the following example:
```python
class YourDatasetProcessor(BlueBERT_GT_Processor):
    
    def __init__(self,
                 use_balanced_neg = False, # if it is Ture, it will randomly selected a subset of negative instances for training
                 max_neg_scale    = 2, # max_neg_scale means how many negative instances will be used for training bert-gt if 'max_neg_scale = 2' means used at most twice negative instances than the other instances
                 balanced_neg     = 'YourNegativeRelationLabel'): # Please replace YourNegativeRelationLabel with your negative instance's label, and balanced_neg will be used only if use_balanced_neg is ture
        super().__init__(
            text_a_idx       = 1, # the index of text column
            in_neighbors_idx = 2, # the index of neighbor column
            label_idx        = 3, # the index of label column
            use_balanced_neg = use_balanced_neg,
            max_neg_scale    = 2,
            balanced_neg     = balanced_neg)
    
    def get_labels(self):
        """See base class."""
        return ["YourNegativeRelationLabel", "YourRelationLabel1", "YourRelationLabel2", "YourRelationLabel3"]
    
    def get_entity_type_dict(self):
        return {'@YourFirstNamedEntityTag$':0, '@YourSecondNamedEntityTag$':1}
        #return {'@YourFirstNamedEntityTag$':0, '@YourSecondNamedEntityTag$':1, '@YourThirdNamedEntityTag$':1}
    
    # if get_entity_type_dict() has more than two NamedEntityTags please remove the following function
    def update_neighbors_by_shortest_path(self, in_neighbors, text_a):
        entity_indices = self.get_entity_indices_by_types(text_a)
        _in_neighbors = in_neighbors.split(' ')
        
        if len(text_a.split(' ')) == len(_in_neighbors):
            new_in_neighbors = graph_algorithm.find_all_shortest_path_between(
                                                  _in_neighbors,
                                                  entity_indices['@YourFirstNamedEntityTag$'],
                                                  entity_indices['@YourSecondNamedEntityTag$'])
        else:
            print('Cannot not process', text_a)
            return in_neighbors

        return ' '.join(new_in_neighbors)
```

In addition,
Please add your YourDatasetProcessor into the processors dict of run_bert_gt.py as followings
```python
    processors = {
        "cdr": CDRGraphProcessor,
        "nary_dgv_mul": NaryDGVMultiProcessor,
        "nary_dv_mul": NaryDVMultiProcessor,
        "nary_dgv_bin": NaryDGVBinaryProcessor,
        "nary_dv_bin": NaryDVBinaryProcessor,
        "ddi": DDIProcessor,
        "chemprot": ChemProtProcessor,
        "tmvar": TmVarGraphProcessor,
        "your_task": YourDatasetProcessor,
    }
```
Then you can run the following command
```
cuda_visible_devices=$cuda_visible_devices python src/run_bert_gt.py \
  --do_train=true \
  --do_eval=false \
  --do_predict=true \
  --task_name="your_task" \
  --vocab_file=biobert_v1.1_pubmed/vocab.txt \
  --bert_config_file=biobert_v1.1_pubmed/bert_config.json \
  --init_checkpoint=biobert_v1.1_pubmed/model.ckpt-1000000 \
  --num_train_epochs=20.0 \
  --max_seq_length=512 \
  --train_batch_size=10 \
  --learning_rate=5e-5 \
  --use_balanced_neg=true \
  --surrounding_words_distance=5 \
  --do_lower_case=false \
  --entity_num=2 \
  --max_num_neighbors=5 \
  --max_num_entity_indices=20 \
  --data_dir=datasets/your_task/processed/ \
  --output_dir=out_your_task_model/
```
You might have to change entity_num if you have more than two NamedEntityTags.

Please let me know if you have any questions/problems. Thank you!

**Comment by juliachen123** [comment]
Hi @ptlai,

I was working on inferencing with custom dataset for nary cross-sentence task as well! I converted the dataset to include neighbors with `DDIProcessor`. When I run the main command with these flags

```  
  --do_train=false\
  --do_eval=false \
  --do_predict=true \
```

It returns an error regarding there is no `label` in the test.tsv. I am wondering if `label` is required when only performing inference?

Thank you in advance!

**Comment by ptlai** [comment]
Hi @julia78118 

A header line is required for the only test.tsv file. For example,

```index	sentence	in_neighbors	label```

And you can add a fake label column.
Please let me know if it doesn't work. Thanks!

**Comment by juliachen123** [comment]
Hi @ptlai ,

Adding a fake label column works. thank you for your response!

**Comment by ArnavHP18** [comment]
> Hello ArnavHP18,
> 
> Sorry for the late response to you. Thank you for your interest in our work.
> To train and test the bert-gt on a custom dataset, there are two steps:
> 
> 1. To convert your dataset into bert-gt's format files, including train.tsv, dev.tsv, and test.tsv. Our codes (in src/dataset_format_converter/) only provide scripts for converting ChemProt/DDI/CDR/n-ary datasets into bert-gt format.
>    So if you can convert your dataset into one of their formats, you can use the corresponding script to convert your dataset into the bert-gt format files.
>    Alternatively, if your dataset cannot convert to one of their formats. You might have to implement your converter to convert your dataset into the bert-gt format file. You should prepare three bert-gt format files (train.tsv, dev.tsv, and test.tsv)
>    You can find the examples of ChemProt/DDI/CDR/n-ary datasets by using our scripts. In general, each file has at least four columns, including 'index,' 'text,' 'neighbors,' 'label.' If you have any problem with the input files, please let me know.
> 2. You have to implement your own YourDatasetProcessor in the run_bert_gt.py as the following example:
> 
> ```python
> class YourDatasetProcessor(BlueBERT_GT_Processor):
>     
>     def __init__(self,
>                  use_balanced_neg = False, # if it is Ture, it will randomly selected a subset of negative instances for training
>                  max_neg_scale    = 2, # max_neg_scale means how many negative instances will be used for training bert-gt if 'max_neg_scale = 2' means used at most twice negative instances than the other instances
>                  balanced_neg     = 'YourNegativeRelationLabel'): # Please replace YourNegativeRelationLabel with your negative instance's label, and balanced_neg will be used only if use_balanced_neg is ture
>         super().__init__(
>             text_a_idx       = 1, # the index of text column
>             in_neighbors_idx = 2, # the index of neighbor column
>             label_idx        = 3, # the index of label column
>             use_balanced_neg = use_balanced_neg,
>             max_neg_scale    = 2,
>             balanced_neg     = balanced_neg)
>     
>     def get_labels(self):
>         """See base class."""
>         return ["YourNegativeRelationLabel", "YourRelationLabel1", "YourRelationLabel2", "YourRelationLabel3"]
>     
>     def get_entity_type_dict(self):
>         return {'@YourFirstNamedEntityTag$':0, '@YourSecondNamedEntityTag$':1}
>         #return {'@YourFirstNamedEntityTag$':0, '@YourSecondNamedEntityTag$':1, '@YourThirdNamedEntityTag$':1}
>     
>     # if get_entity_type_dict() has more than two NamedEntityTags please remove the following function
>     def update_neighbors_by_shortest_path(self, in_neighbors, text_a):
>         entity_indices = self.get_entity_indices_by_types(text_a)
>         _in_neighbors = in_neighbors.split(' ')
>         
>         if len(text_a.split(' ')) == len(_in_neighbors):
>             new_in_neighbors = graph_algorithm.find_all_shortest_path_between(
>                                                   _in_neighbors,
>                                                   entity_indices['@YourFirstNamedEntityTag$'],
>                                                   entity_indices['@YourSecondNamedEntityTag$'])
>         else:
>             print('Cannot not process', text_a)
>             return in_neighbors
> 
>         return ' '.join(new_in_neighbors)
> ```
> 
> In addition,
> Please add your YourDatasetProcessor into the processors dict of run_bert_gt.py as followings
> 
> ```python
>     processors = {
>         "cdr": CDRGraphProcessor,
>         "nary_dgv_mul": NaryDGVMultiProcessor,
>         "nary_dv_mul": NaryDVMultiProcessor,
>         "nary_dgv_bin": NaryDGVBinaryProcessor,
>         "nary_dv_bin": NaryDVBinaryProcessor,
>         "ddi": DDIProcessor,
>         "chemprot": ChemProtProcessor,
>         "tmvar": TmVarGraphProcessor,
>         "your_task": YourDatasetProcessor,
>     }
> ```
> 
> Then you can run the following command
> 
> ```
> cuda_visible_devices=$cuda_visible_devices python src/run_bert_gt.py \
>   --do_train=true \
>   --do_eval=false \
>   --do_predict=true \
>   --task_name="your_task" \
>   --vocab_file=biobert_v1.1_pubmed/vocab.txt \
>   --bert_config_file=biobert_v1.1_pubmed/bert_config.json \
>   --init_checkpoint=biobert_v1.1_pubmed/model.ckpt-1000000 \
>   --num_train_epochs=20.0 \
>   --max_seq_length=512 \
>   --train_batch_size=10 \
>   --learning_rate=5e-5 \
>   --use_balanced_neg=true \
>   --surrounding_words_distance=5 \
>   --do_lower_case=false \
>   --entity_num=2 \
>   --max_num_neighbors=5 \
>   --max_num_entity_indices=20 \
>   --data_dir=datasets/your_task/processed/ \
>   --output_dir=out_your_task_model/
> ```
> 
> You might have to change entity_num if you have more than two NamedEntityTags.
> 
> Please let me know if you have any questions/problems. Thank you!

**Comment by ArnavHP18** [comment]
Thank you.
